src/metabase/automagic_dashboards/core.clj

```diff
@@ -219,22 +345,20 @@
   (filter #(-> % :entity_type (isa? tablespec)) tables))
 
 (defn- fill-templates
-  [template-type context bindings form]
-  (walk/postwalk
-   (fn [form]
-     (if (string? form)
-       (str/replace form #"\[\[(\w+)\]\]"
-                    (fn [[_ identifier]]
-                      (->reference template-type (or (-> identifier
-                                                         ((merge {"this" (-> context :root :entity)}
-                                                                 bindings)))
-                                                     (-> identifier
-                                                         rules/->entity
-                                                         (filter-tables (:tables context))
-                                                         first)
-                                                     identifier))))
-       form))
-   form))
+  [template-type {:keys [root tables]} bindings form]
+  (let [bindings (some-fn (merge {"this" (-> root
+                                             :entity
+                                             (assoc :full-name (:full-name root)))}
+                                 bindings)
+                          (comp first #(filter-tables % tables) rules/->entity)
+                          identity)]
+    (walk/postwalk
+     (fn [form]
+       (if (string? form)
+         (str/replace form #"\[\[(\w+)\]\]" (fn [[_ identifier]]
```

Comment:
I've got to admit, working on this query structure directly as un-expanded and separate from the code in the query-processor makes me pretty nervous. We don't have great tools right now to work on the structure, and picking through the structures in this way seems pretty tedious and error prone. I think this will get tougher as we evolve our language. This is one example, but there are lots of this kind of tree manipulation done in this code.

Reply:
I agree. I'll do a bigger refactor once the main pieces are in place and try to add to/unify with qp as much as I can. Some tree walking is inevitable but a cleaner approach would be to do it once and assemble some flatter data structure that would then be queried.